File: web/src/features/batch/medicationNameMatcher.ts
/**
 * Pure-function deterministic medication-name resolver.
 *
 * Mirrors `MedicationNameMatcher.kt` (Android) and
 * `backend/app/services/medication_name_matcher.py` byte-for-byte: the same
 * command + medication list MUST produce the same `MatchResult` on every
 * surface, otherwise the audit's failure-mode #6
 * (case/whitespace/unicode brittleness) reopens.
 *
 * NEVER does fuzzy or substring matching. Typos and partial words return
 * `NoMatch` so the caller falls through to the Haiku batch path. This is
 * the audit's failure-mode #2 firewall: silent typo confidence is exactly
 * the bug we are guarding against, so a stricter "exact whole-word match
 * or nothing" contract is the safe default.
 *
 * Normalization (must stay byte-identical with Kotlin + Python twins):
 *   1. Unicode NFC normalize
 *   2. trim
 *   3. lowercase
 *   4. strip trailing ASCII punctuation [.,!?;:]
 */

export interface Medication {
  id: string;
  name: string;
  display_label?: string | null;
}

export interface AmbiguousPhrase {
  phrase: string;
  candidate_entity_ids: string[];
}

export type MatchResult =
  | { kind: 'no_match' }
  | { kind: 'unambiguous'; matches: Record<string, string> }
  | { kind: 'ambiguous'; phrases: AmbiguousPhrase[] }
  | {
      kind: 'mixed';
      unambiguous: Record<string, string>;
      ambiguous: AmbiguousPhrase[];
    };

const TRAILING_PUNCT = new Set(['.', ',', '!', '?', ';', ':']);

export function normalize(s: string): string {
  const nfc = s.normalize('NFC');
  const trimmed = nfc.trim();
  const lower = trimmed.toLowerCase();
  let end = lower.length;
  while (end > 0 && TRAILING_PUNCT.has(lower[end - 1])) end -= 1;
  return lower.slice(0, end);
}

function isWordChar(ch: string): boolean {
  if (!ch) return false;
  return /[\p{L}\p{N}_]/u.test(ch);
}

function buildKeyIndex(medications: Medication[]): Map<string, string[]> {
  const keyToIds = new Map<string, string[]>();
  for (const med of medications) {
    for (const raw of [med.name, med.display_label ?? null]) {
      if (raw == null) continue;
      const key = normalize(raw);
      if (!key) continue;
      const ids = keyToIds.get(key) ?? [];
      if (!ids.includes(med.id)) ids.push(med.id);
      keyToIds.set(key, ids);
    }
  }
  return keyToIds;
}

function findLongestKeyAt(
  haystack: string,
  start: number,
  keys: string[],
): string | null {
  for (const k of keys) {
    const end = start + k.length;
    if (end > haystack.length) continue;
    if (end < haystack.length && isWordChar(haystack[end])) continue;
    if (haystack.slice(start, end) === k) return k;
  }
  return null;
}

function classify(
  unambiguous: Record<string, string>,
  ambiguousMap: Map<string, string[]>,
): MatchResult {
  const ambiguous: AmbiguousPhrase[] = [];
  for (const [phrase, ids] of ambiguousMap) {
    ambiguous.push({
      phrase,
      candidate_entity_ids: [...ids].sort(),
    });
  }
  const hasUnambiguous = Object.keys(unambiguous).length > 0;
  const hasAmbiguous = ambiguous.length > 0;
  if (!hasUnambiguous && !hasAmbiguous) return { kind: 'no_match' };
  if (!hasAmbiguous) return { kind: 'unambiguous', matches: unambiguous };
  if (!hasUnambiguous) return { kind: 'ambiguous', phrases: ambiguous };
  return { kind: 'mixed', unambiguous, ambiguous };
}

export function matchMedicationsInCommand(
  commandText: string,
  medications: Medication[],
): MatchResult {
  if (!commandText || !commandText.trim() || medications.length === 0) {
    return { kind: 'no_match' };
  }
  const normalizedCommand = normalize(commandText);
  if (!normalizedCommand) return { kind: 'no_match' };

  const keyToIds = buildKeyIndex(medications);
  if (keyToIds.size === 0) return { kind: 'no_match' };

  const keys = [...keyToIds.keys()].sort((a, b) => b.length - a.length);
  const unambiguous: Record<string, string> = {};
  const ambiguousMap = new Map<string, string[]>();

  let i = 0;
  while (i < normalizedCommand.length) {
    const atWordStart =
      isWordChar(normalizedCommand[i]) &&
      (i === 0 || !isWordChar(normalizedCommand[i - 1]));
    if (!atWordStart) {
      i += 1;
      continue;
    }
    const matchedKey = findLongestKeyAt(normalizedCommand, i, keys);
    if (matchedKey != null) {
      const ids = keyToIds.get(matchedKey)!;
      if (ids.length === 1) {
        unambiguous[matchedKey] = ids[0];
      } else {
        const bucket = ambiguousMap.get(matchedKey) ?? [];
        for (const medId of ids) {
          if (!bucket.includes(medId)) bucket.push(medId);
        }
        ambiguousMap.set(matchedKey, bucket);
      }
      i += matchedKey.length;
    } else {
      while (
        i < normalizedCommand.length &&
        isWordChar(normalizedCommand[i])
      ) {
        i += 1;
      }
    }
  }

  return classify(unambiguous, ambiguousMap);
}

File: backend/app/services/medication_name_matcher.py
"""Pure-function deterministic medication-name resolver.

Mirrors `MedicationNameMatcher.kt` (Android) and
`web/src/features/batch/medicationNameMatcher.ts` byte-for-byte: the same
command + medication list MUST produce the same `MatchResult` on every
surface, otherwise the audit's failure-mode #6 (case/whitespace/unicode
brittleness) reopens.

NEVER does fuzzy or substring matching. Typos and partial words return
`NoMatch` so the caller falls through to the Haiku batch path. This is the
audit's failure-mode #2 firewall: silent typo confidence is exactly the bug
we are guarding against, so a stricter "exact whole-word match or nothing"
contract is the safe default.

Normalization (must stay byte-identical with Kotlin + TS twins):
    1. Unicode NFC normalize
    2. strip
    3. lowercase
    4. strip trailing ASCII punctuation [.,!?;:]
"""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass
from typing import Iterable, Optional, Union

_TRAILING_PUNCT = ".,!?;:"


@dataclass(frozen=True)
class Medication:
    id: str
    name: str
    display_label: Optional[str] = None


@dataclass(frozen=True)
class AmbiguousPhrase:
    phrase: str
    candidate_entity_ids: list[str]


@dataclass(frozen=True)
class NoMatch:
    pass


@dataclass(frozen=True)
class Unambiguous:
    """`matches` is keyed by the normalized phrase that hit the medication."""

    matches: dict[str, str]


@dataclass(frozen=True)
class Ambiguous:
    phrases: list[AmbiguousPhrase]


@dataclass(frozen=True)
class Mixed:
    unambiguous: dict[str, str]
    ambiguous: list[AmbiguousPhrase]


MatchResult = Union[NoMatch, Unambiguous, Ambiguous, Mixed]


def normalize(s: str) -> str:
    """Apply the four-step normalization. Must match the Kotlin + TS twins."""

    nfc = unicodedata.normalize("NFC", s)
    trimmed = nfc.strip()
    lower = trimmed.lower()
    end = len(lower)
    while end > 0 and lower[end - 1] in _TRAILING_PUNCT:
        end -= 1
    return lower[:end]


def _is_word_char(ch: str) -> bool:
    return ch.isalnum() or ch == "_"


def _build_key_index(medications: Iterable[Medication]) -> dict[str, list[str]]:
    """Map normalized name/label -> ordered list of medication ids that share it.

    The list preserves insertion order so two meds named "Wellbutrin" come back
    in a predictable order (matches the Kotlin + TS twins).
    """

    key_to_ids: dict[str, list[str]] = {}
    for med in medications:
        for raw in (med.name, med.display_label):
            if raw is None:
                continue
            key = normalize(raw)
            if not key:
                continue
            ids = key_to_ids.setdefault(key, [])
            if med.id not in ids:
                ids.append(med.id)
    return key_to_ids


def _find_longest_key_at(haystack: str, start: int, keys: list[str]) -> Optional[str]:
    for k in keys:
        end = start + len(k)
        if end > len(haystack):
            continue
        if end < len(haystack) and _is_word_char(haystack[end]):
            continue
        if haystack[start:end] == k:
            return k
    return None


def _classify(
    unambiguous: dict[str, str],
    ambiguous_map: dict[str, list[str]],
) -> MatchResult:
    ambiguous_list = [
        AmbiguousPhrase(phrase=phrase, candidate_entity_ids=sorted(ids))
        for phrase, ids in ambiguous_map.items()
    ]
    if not unambiguous and not ambiguous_list:
        return NoMatch()
    if not ambiguous_list:
        return Unambiguous(matches=unambiguous)
    if not unambiguous:
        return Ambiguous(phrases=ambiguous_list)
    return Mixed(unambiguous=unambiguous, ambiguous=ambiguous_list)


def match(command_text: str, medications: list[Medication]) -> MatchResult:
    if not command_text or not command_text.strip() or not medications:
        return NoMatch()
    normalized_command = normalize(command_text)
    if not normalized_command:
        return NoMatch()

    key_to_ids = _build_key_index(medications)
    if not key_to_ids:
        return NoMatch()

    keys = sorted(key_to_ids.keys(), key=len, reverse=True)
    unambiguous: dict[str, str] = {}
    ambiguous_map: dict[str, list[str]] = {}

    i = 0
    while i < len(normalized_command):
        at_word_start = _is_word_char(normalized_command[i]) and (
            i == 0 or not _is_word_char(normalized_command[i - 1])
        )
        if not at_word_start:
            i += 1
            continue
        matched_key = _find_longest_key_at(normalized_command, i, keys)
        if matched_key is not None:
            ids = key_to_ids[matched_key]
            if len(ids) == 1:
                unambiguous[matched_key] = ids[0]
            else:
                bucket = ambiguous_map.setdefault(matched_key, [])
                for med_id in ids:
                    if med_id not in bucket:
                        bucket.append(med_id)
            i += len(matched_key)
        else:
            while i < len(normalized_command) and _is_word_char(normalized_command[i]):
                i += 1

    return _classify(unambiguous, ambiguous_map)
